pkg/kubectl/cmd/create_role.go

```diff
@@ -0,0 +1,255 @@
+		},
+	}
+	cmdutil.AddApplyAnnotationFlags(cmd)
+	cmdutil.AddValidateFlags(cmd)
+	cmdutil.AddPrinterFlags(cmd)
+	cmdutil.AddDryRunFlag(cmd)
+	cmd.Flags().StringSliceVar(&c.Verbs, "verb", []string{}, "verb that applies to the resources contained in the rule")
+	cmd.Flags().StringSlice("resource", []string{}, "resource that the rule applies to")
+	cmd.Flags().StringSliceVar(&c.ResourceNames, "resource-name", []string{}, "resource in the white list that the rule applies to")
+
+	return cmd
+}
+
+func (c *CreateRoleOptions) Complete(cmd *cobra.Command, args []string) error {
+	name, err := NameFromCommandArgs(cmd, args)
+	if err != nil {
+		return err
+	}
+	c.Name = name
+
+	// Support resource.group pattern. If no API Group specified, use "" as core API Group.
+	// e.g. --resource=pods,deployments.extensions
+	resources := cmdutil.GetFlagStringSlice(cmd, "resource")
+	for _, r := range resources {
+		sections := strings.Split(r, ".")
+		resourceName, apiGroup := "", ""
+
+		if len(sections) == 1 {
+			resourceName = r
+		} else {
+			resourceName, apiGroup = sections[0], strings.Join(sections[1:], ".")
+		}
+
+		c.Resources = append(c.Resources, schema.GroupVersionResource{Group: apiGroup, Resource: resourceName})
+	}
+
+	return nil
+}
+
+func (c *CreateRoleOptions) Validate(f cmdutil.Factory) error {
+	if c.Name == "" {
+		return fmt.Errorf("name must be specified")
+	}
+
+	// validate verbs.
+	if len(c.Verbs) == 0 {
+		return fmt.Errorf("at least one verb must be specified")
+	}
+
+	for _, v := range c.Verbs {
+		if !arrayContains(validResourceVerbs, v) && !arrayContains(validNonResourceVerbs, v) {
+			return fmt.Errorf("invalid verb: '%s'", v)
+		}
+	}
+
+	// validate resources.
+	mapper, _ := f.Object()
+
+	if len(c.Resources) == 0 {
+		return fmt.Errorf("at least one resource must be specified")
+	}
+
+	for _, r := range c.Resources {
+		_, err := mapper.ResourceFor(r)
+		if err != nil {
+			return err
+		}
+	}
+
+	// validate resource names, can not apply resource names to multiple resources.
+	if len(c.ResourceNames) > 0 && len(c.Resources) > 1 {
+		return fmt.Errorf("resource name(s) can not be applied to multiple resources")
+	}
+
+	return nil
+}
+
+func (c *CreateRoleOptions) RunCreateRole(f cmdutil.Factory, cmdOut io.Writer, cmd *cobra.Command, args []string) error {
+	mapper, typer := f.Object()
+	dryRun, outputFormat := cmdutil.GetDryRunFlag(cmd), cmdutil.GetFlagString(cmd, "output")
+
+	// Remove duplicate verbs.
+	verbs, containsVerbAll := []string{}, false
+	for _, v := range c.Verbs {
+		if !arrayContains(verbs, v) {
+			verbs = append(verbs, v)
+		}
+		if v == "*" {
+			containsVerbAll = true
+		}
+	}
+	// VerbAll respresents all kinds of verbs.
+	if containsVerbAll {
+		verbs = []string{"*"}
+	}
+
+	// Remove duplicate resources and transfer resource name from short form to long form.
+	groupResourceMapping := map[string][]string{}
+	for _, r := range c.Resources {
+		resource, err := mapper.ResourceFor(r)
+		if err != nil {
```

Comment:
document that this prevents pointing to non-existent resources and allows resource shortnames to be used in the CLI.